File: frontend/src/types/instruction-set.ts
/**
 * 指令集相关的TypeScript类型定义
 */

// 枚举类型
export enum InstructionSetStatus {
  DRAFT = 'DRAFT',
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  ARCHIVED = 'ARCHIVED'
}

export enum NodeType {
  EXECUTOR = 'EXECUTOR',      // 执行器节点（根节点）
  CONDITION = 'CONDITION',    // 条件判断节点
  ACTION = 'ACTION',          // 动作执行节点
  BRANCH = 'BRANCH',          // 分支节点
  AGGREGATOR = 'AGGREGATOR',  // 聚合器节点
  CLASSIFIER = 'CLASSIFIER',  // 分类器节点
  RESULT = 'RESULT'           // 结果节点
}

export enum ConditionType {
  CONTAINS = 'CONTAINS',
  REGEX = 'REGEX',
  LENGTH = 'LENGTH',
  KEYWORD = 'KEYWORD',
  AI_CLASSIFICATION = 'AI_CLASSIFICATION',  // AI分类判断
  SENTIMENT_ANALYSIS = 'SENTIMENT_ANALYSIS', // 情感分析
  CONTENT_SAFETY = 'CONTENT_SAFETY',        // 内容安全检测
  CUSTOM_FUNCTION = 'CUSTOM_FUNCTION'       // 自定义函数
}

export enum ActionType {
  RETURN_TEXT = 'RETURN_TEXT',
  EXTRACT_INFO = 'EXTRACT_INFO',
  TRANSFORM = 'TRANSFORM',
  CLASSIFY = 'CLASSIFY',
  APPROVE = 'APPROVE',          // 通过审核
  REJECT = 'REJECT',            // 拒绝审核
  FLAG_CONTENT = 'FLAG_CONTENT', // 标记内容
  SEND_NOTIFICATION = 'SEND_NOTIFICATION', // 发送通知
  LOG_EVENT = 'LOG_EVENT',      // 记录事件
  CUSTOM_ACTION = 'CUSTOM_ACTION' // 自定义动作
}

export enum ExecutionStatus {
  SUCCESS = 'SUCCESS',
  FAILED = 'FAILED',
  TIMEOUT = 'TIMEOUT',
  PENDING = 'PENDING',
  CANCELLED = 'CANCELLED'
}

// 新增枚举类型
export enum LogicOperator {
  AND = 'AND',
  OR = 'OR',
  NOT = 'NOT'
}

export enum ExecutionStrategy {
  SEQUENTIAL = 'SEQUENTIAL',    // 顺序执行
  PARALLEL = 'PARALLEL',       // 并行执行
  CONDITIONAL = 'CONDITIONAL',  // 条件执行
  EARLY_STOP = 'EARLY_STOP'    // 早停执行
}

export enum ScoreType {
  ACCURACY = 'ACCURACY',        // 准确性
  CONFIDENCE = 'CONFIDENCE',    // 置信度
  PERFORMANCE = 'PERFORMANCE',  // 性能
  RELEVANCE = 'RELEVANCE',     // 相关性
  SAFETY = 'SAFETY'            // 安全性
}

export enum AggregationMethod {
  WEIGHTED_AVERAGE = 'WEIGHTED_AVERAGE',
  MAX_SCORE = 'MAX_SCORE',
  MIN_SCORE = 'MIN_SCORE',
  MAJORITY_VOTE = 'MAJORITY_VOTE',
  CONSENSUS = 'CONSENSUS'
}

export enum RiskLevel {
  SAFE = 'safe',
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical'
}

// 基础接口
export interface InstructionSet {
  id: number;
  name: string;
  description?: string;
  version: string;
  status: InstructionSetStatus;
  created_by: number;
  created_at: string;
  updated_at: string;
  is_deleted: boolean;
}

export interface InstructionSetCreate {
  name: string;
  description?: string;
  version?: string;
  status?: InstructionSetStatus;
  created_by: number;
}

export interface InstructionSetUpdate {
  name?: string;
  description?: string;
  version?: string;
  status?: InstructionSetStatus;
}

export interface InstructionNode {
  id: number;
  instruction_set_id: number;
  parent_id?: number;
  title: string;
  description?: string;
  node_type: NodeType;
  condition_text?: string;      // 条件文本
  condition_type: ConditionType; // 条件类型
  keywords?: string;           // 关键词列表（逗号分隔）
  risk_level: RiskLevel;       // 风险等级
  metadata?: Record<string, any>; // 扩展元数据
  sort_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  children?: InstructionNode[]; // 子节点列表
}

export interface InstructionNodeCreate {
  instruction_set_id?: number;
  parent_id?: number;
  title: string;
  description?: string;
  node_type: NodeType;
  condition_text?: string;
  condition_type?: ConditionType;
  keywords?: string;
  risk_level?: RiskLevel;
  metadata?: Record<string, any>;
  sort_order?: number;
  is_active?: boolean;
}

export interface InstructionNodeUpdate {
  parent_id?: number;
  title?: string;
  description?: string;
  node_type?: NodeType;
  condition_text?: string;
  condition_type?: ConditionType;
  keywords?: string;
  risk_level?: RiskLevel;
  metadata?: Record<string, any>;
  sort_order?: number;
  is_active?: boolean;
}

export interface InstructionExecution {
  id: number;
  instruction_set_id: number;
  input_text: string;
  execution_path?: string;
  final_result?: string;
  confidence_score?: number;
  execution_time_ms?: number;
  status: ExecutionStatus;
  error_message?: string;
  created_at: string;
}

export interface InstructionExecutionCreate {
  instruction_set_id: number;
  input_text: string;
  execution_path?: string;
  final_result?: string;
  confidence_score?: number;
  execution_time_ms?: number;
  status: ExecutionStatus;
  error_message?: string;
}

// 执行请求和响应
export interface InstructionExecuteRequest {
  instruction_set_id: number;
  input_text: string;
  save_execution?: boolean;
}

export interface InstructionExecuteResponse {
  execution_path: ExecutionPathItem[];
  final_result: string;
  confidence_score: number;
  execution_time_ms: number;
  metadata?: Record<string, any>;
  execution_id?: number;
}

export interface ExecutionPathItem {
  node_id: number;
  node_title: string;
  node_type: string;
  timestamp: number;
}

// 树形结构
export interface InstructionTreeNode {
  id: number;
  title: string;
  node_type: NodeType;
  parent_id?: number;
  keywords?: string;
  condition_text?: string;
  risk_level?: string;
  sort_order: number;
  is_active: boolean;
  node_number?: string;
  children: InstructionTreeNode[];
}

// 条件配置
export interface ConditionConfig {
  type: ConditionType;
  value: string;
  operator?: 'eq' | 'gt' | 'lt' | 'gte' | 'lte';
}

// 动作配置
export interface ActionConfig {
  type: ActionType;
  value: string;
  pattern?: string;
  transform_type?: 'upper' | 'lower' | 'title' | 'reverse';
  categories?: string[];
  keywords_map?: Record<string, string[]>;
}

// 统计信息
export interface InstructionSetStatistics {
  total_executions: number;
  success_executions: number;
  failed_executions: number;
  average_execution_time: number;
  average_confidence_score: number;
  most_used_paths: Array<{
    path: string;
    count: number;
  }>;
}

// API响应包装
export interface ApiResponse<T> {
  data: T;
  message?: string;
  success: boolean;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface InstructionSetListResponse {
  success: boolean;
  data: InstructionSet[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  has_next: boolean;
  has_prev: boolean;
}

// 查询参数
export interface InstructionSetQuery {
  skip?: number;
  limit?: number;
  status?: InstructionSetStatus;
  created_by?: number;
  search?: string;
}

export interface InstructionNodeQuery {
  instruction_set_id: number;
  parent_id?: number;
  skip?: number;
  limit?: number;
}

export interface InstructionExecutionQuery {
  instruction_set_id: number;
  skip?: number;
  limit?: number;
  status?: ExecutionStatus;
  start_date?: string;
  end_date?: string;
}

// 表单数据
export interface InstructionSetFormData {
  name: string;
  description: string;
  version: string;
  status: InstructionSetStatus;
}

export interface InstructionNodeFormData {
  title: string;
  description: string;
  node_type: NodeType;
  condition_text: string;
  condition_type: ConditionType;
  keywords: string[];
  metadata: Record<string, any>;
  sort_order: number;
  is_active: boolean;
}

// 树操作
export interface TreeOperation {
  type: 'add' | 'edit' | 'delete' | 'move';
  node?: InstructionTreeNode;
  parent_id?: number;
  new_parent_id?: number;
  sort_order?: number;
}

// 拖拽数据
export interface DragDropData {
  dragNode: InstructionTreeNode;
  dropNode: InstructionTreeNode;
  dropPosition: 'before' | 'after' | 'inside';
}

// 新增配置接口

/**
 * 执行器配置接口
 */
export interface ExecutorConfig {
  strategy: ExecutionStrategy;
  max_parallel_nodes?: number;
  timeout_ms?: number;
  retry_count?: number;
  early_stop_conditions?: EarlyStopCondition[];
  global_variables?: Record<string, any>;
}

/**
 * 早停条件配置
 */
export interface EarlyStopCondition {
  condition_type: 'confidence_threshold' | 'time_limit' | 'error_rate';
  threshold_value: number;
  operator: 'gt' | 'lt' | 'eq' | 'gte' | 'lte';
}

/**
 * 评分配置接口
 */
export interface ScoreConfig {
  enabled: boolean;
  score_types: ScoreType[];
  weights?: Record<ScoreType, number>;
  aggregation_method: AggregationMethod;
  min_threshold?: number;
  max_threshold?: number;
}

/**
 * 复杂条件配置接口
 */
export interface ComplexConditionConfig {
  logic_operator: LogicOperator;
  conditions: ConditionItem[];
  nested_groups?: ComplexConditionConfig[];
}

/**
 * 条件项接口
 */
export interface ConditionItem {
  id: string;
  type: ConditionType;
  field: string;
  operator: 'eq' | 'ne' | 'gt' | 'lt' | 'gte' | 'lte' | 'contains' | 'regex';
  value: any;
  weight?: number;
  description?: string;
}

/**
 * 聚合器配置接口
 */
export interface AggregatorConfig {
  input_nodes: number[];
  aggregation_method: AggregationMethod;
  weights?: Record<number, number>;
  conflict_resolution: 'highest_score' | 'majority_vote' | 'weighted_average';
  min_consensus_threshold?: number;
}

/**
 * 分类器配置接口
 */
export interface ClassifierConfig {
  categories: ClassificationCategory[];
  default_category?: string;
  confidence_threshold: number;
  multi_label: boolean;
  model_config?: {
    model_name: string;
    api_endpoint?: string;
    parameters?: Record<string, any>;
  };
}

/**
 * 分类类别接口
 */
export interface ClassificationCategory {
  id: string;
  name: string;
  description?: string;
  keywords?: string[];
  patterns?: string[];
  examples?: string[];
  weight?: number;
}

/**
 * 执行结果接口
 */
export interface NodeExecutionResult {
  node_id: number;
  status: ExecutionStatus;
  result: any;
  scores?: Record<ScoreType, number>;
  execution_time_ms: number;
  error_message?: string;
  metadata?: Record<string, any>;
}

/**
 * 指令集执行上下文
 */
export interface ExecutionContext {
  input_data: any;
  variables: Record<string, any>;
  execution_path: number[];
  current_node_id?: number;
  start_time: number;
  timeout_ms?: number;
  user_id?: number;
  session_id?: string;
}

File: backend/app/schemas/instruction_set.py
from typing import List, Optional, Dict, Any
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field
from app.models.instruction_set import (
    InstructionSetStatus, NodeType, ConditionType, ActionType, ExecutionStatus, RiskLevel
)


class InstructionSetBase(BaseModel):
    """指令集基础模型"""
    name: str = Field(..., description="指令集名称")
    description: Optional[str] = Field(None, description="指令集描述")
    version: str = Field(default="1.0.0", description="版本号")
    status: InstructionSetStatus = Field(default=InstructionSetStatus.DRAFT, description="状态")


class InstructionSetCreate(InstructionSetBase):
    """创建指令集请求模型"""
    created_by: Optional[int] = Field(None, description="创建者ID")


class InstructionSetUpdate(BaseModel):
    """更新指令集请求模型"""
    name: Optional[str] = Field(None, description="指令集名称")
    description: Optional[str] = Field(None, description="指令集描述")
    version: Optional[str] = Field(None, description="版本号")
    status: Optional[InstructionSetStatus] = Field(None, description="状态")


class InstructionNodeBase(BaseModel):
    """指令节点基础模型"""
    parent_id: Optional[int] = Field(None, description="父节点ID")
    node_type: NodeType = Field(..., description="节点类型")
    title: str = Field(..., description="节点标题")
    description: Optional[str] = Field(None, description="节点描述")
    keywords: Optional[str] = Field(None, description="关键词")
    condition_text: Optional[str] = Field(None, description="条件文本")
    condition_type: ConditionType = Field(default=ConditionType.AI_CLASSIFICATION, description="条件类型")
    action_type: ActionType = Field(default=ActionType.CONTINUE, description="动作类型")
    result_value: Optional[str] = Field(None, description="结果值")
    result_confidence: Decimal = Field(default=Decimal('0.00'), description="结果置信度")
    risk_level: RiskLevel = Field(default=RiskLevel.medium, description="风险等级")
    sort_order: int = Field(default=0, description="排序顺序")
    is_active: bool = Field(default=True, description="是否激活")
    meta_data: Optional[Dict[str, Any]] = Field(None, description="扩展元数据", alias="metadata")


class InstructionNodeCreate(InstructionNodeBase):
    """创建指令节点请求模型"""
    instruction_set_id: int = Field(..., description="所属指令集ID")


class InstructionNodeUpdate(BaseModel):
    """更新指令节点请求模型"""
    parent_id: Optional[int] = Field(None, description="父节点ID")
    node_type: Optional[NodeType] = Field(None, description="节点类型")
    title: Optional[str] = Field(None, description="节点标题")
    description: Optional[str] = Field(None, description="节点描述")
    keywords: Optional[str] = Field(None, description="关键词")
    condition_text: Optional[str] = Field(None, description="条件文本")
    condition_type: Optional[ConditionType] = Field(None, description="条件类型")
    action_type: Optional[ActionType] = Field(None, description="动作类型")
    result_value: Optional[str] = Field(None, description="结果值")
    result_confidence: Optional[Decimal] = Field(None, description="结果置信度")
    risk_level: Optional[RiskLevel] = Field(None, description="风险等级")
    sort_order: Optional[int] = Field(None, description="排序顺序")
    is_active: Optional[bool] = Field(None, description="是否激活")
    meta_data: Optional[Dict[str, Any]] = Field(None, description="扩展元数据", alias="metadata")


class InstructionNode(InstructionNodeBase):
    """指令节点响应模型"""
    id: int = Field(..., description="节点ID")
    instruction_set_id: int = Field(..., description="所属指令集ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    children: List['InstructionNode'] = Field(default=[], description="子节点列表")

    class Config:
        from_attributes = True


class InstructionSet(InstructionSetBase):
    """指令集响应模型"""
    id: int = Field(..., description="指令集ID")
    created_by: Optional[int] = Field(None, description="创建者ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


class InstructionSetWithNodes(InstructionSet):
    """包含节点的指令集响应模型"""
    nodes: List[InstructionNode] = Field(default=[], description="节点列表")
    executions: List['InstructionExecution'] = Field(default=[], description="执行记录列表")

    class Config:
        from_attributes = True


class InstructionExecutionBase(BaseModel):
    """指令执行基础模型"""
    input_text: str = Field(..., description="输入文本")
    execution_path: Optional[List[int]] = Field(None, description="执行路径")
    final_result: Optional[str] = Field(None, description="最终结果")
    confidence_score: Optional[Decimal] = Field(None, description="置信度分数")
    execution_time_ms: Optional[int] = Field(None, description="执行时间（毫秒）")
    status: ExecutionStatus = Field(default=ExecutionStatus.SUCCESS, description="执行状态")
    error_message: Optional[str] = Field(None, description="错误信息")


class InstructionExecutionCreate(InstructionExecutionBase):
    """创建指令执行请求模型"""
    instruction_set_id: int = Field(..., description="指令集ID")


class InstructionExecution(InstructionExecutionBase):
    """指令执行响应模型"""
    id: int = Field(..., description="执行ID")
    instruction_set_id: int = Field(..., description="指令集ID")
    created_at: datetime = Field(..., description="创建时间")

    class Config:
        from_attributes = True


class InstructionExecuteRequest(BaseModel):
    """执行指令集请求模型"""
    instruction_set_id: int = Field(..., description="指令集ID")
    input_text: str = Field(..., description="输入文本")
    save_execution: bool = Field(default=True, description="是否保存执行记录")


class InstructionExecuteResponse(BaseModel):
    """执行指令集响应模型"""
    final_result: str = Field(..., description="最终结果")
    confidence_score: Decimal = Field(..., description="置信度分数")
    execution_path: List[int] = Field(..., description="执行路径")
    execution_time_ms: int = Field(..., description="执行时间（毫秒）")
    execution_id: Optional[int] = Field(None, description="执行记录ID")


class InstructionTreeNode(BaseModel):
    """指令树节点模型"""
    id: int = Field(..., description="节点ID")
    title: str = Field(..., description="节点标题")
    node_type: NodeType = Field(..., description="节点类型")
    parent_id: Optional[int] = Field(None, description="父节点ID")
    keywords: Optional[str] = Field(None, description="关键词")
    condition_text: Optional[str] = Field(None, description="条件文本")
    risk_level: RiskLevel = Field(default=RiskLevel.medium, description="风险等级")
    sort_order: int = Field(..., description="排序顺序")
    is_active: bool = Field(..., description="是否激活")
    node_number: str = Field(..., description="节点编号")
    children: List['InstructionTreeNode'] = Field(default=[], description="子节点")


class PaginatedResponse(BaseModel):
    """分页响应模型"""
    success: bool = Field(default=True, description="请求是否成功")
    data: List[Any] = Field(..., description="数据列表")
    total: int = Field(..., description="总记录数")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页大小")
    total_pages: int = Field(..., description="总页数")
    has_next: bool = Field(..., description="是否有下一页")
    has_prev: bool = Field(..., description="是否有上一页")


class InstructionSetListResponse(BaseModel):
    """指令集列表响应模型"""
    success: bool = Field(default=True, description="请求是否成功")
    data: List[InstructionSet] = Field(..., description="指令集列表")
    total: int = Field(..., description="总记录数")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页大小")
    total_pages: int = Field(..., description="总页数")
    has_next: bool = Field(..., description="是否有下一页")
    has_prev: bool = Field(..., description="是否有上一页")


# 解决前向引用问题
InstructionNode.model_rebuild()
InstructionTreeNode.model_rebuild()